src/hotspot/share/runtime/mutexLocker.cpp

```diff
@@ -409,3 +422,36 @@ void print_owned_locks_on_error(outputStream* st) {
   }
   if (none) st->print_cr("None");
 }
+
+void print_lock_ranks(outputStream* st) {
+  st->print_cr("VM Mutex/Monitor ranks: ");
+
+#ifdef ASSERT
+  int min_rank = INT_MAX;
+  int max_rank = INT_MIN;
+
+  for (int i = 0; i < _num_mutex; i++) {
+    Mutex *m = _mutex_array[i];
+    int r = (int) m->rank();
+    if (min_rank > r) min_rank = r;
+    if (max_rank < r) max_rank = r;
+  }
+
+  for (int r = min_rank; r <= max_rank; r++) {
```

Comment:
Ah, I was being defensive about this: computing min/max without relying on the actual `Rank` definition. While `0` is plausibly the lowest rank, I don't really want to rely on `safepoint` being the last rank. This is debugging code, so we can spend some time on this. Makes sense?

Reply:
All the other code in this file thinks safepoint is the last rank (it's an invariant) and INT_MAX is a bit large even for debugging code.  I think safepoint would be better. I marked it as reviewed so am not really insistent., but it looked weird to me.

Reply:
Yeah, I can now see the `Mutex` constructor asserts the rank is within (0...safepoint). But it still feels a bit fragile to rely on, so I left it as is. This debugging code, we better off overdoing stuff rather than misleading developers. I added a few comments, though!